x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1641,6 +1645,40 @@ public boolean test(String profile, InetSocketAddress peerAddress) {
                 sslConfiguration = null;
                 populateClientCertificate = (channel, threadContext) -> {};
             }
+            final AuthenticationService authenticationService = this.authcService.get();
+            final ThreadContext threadContext = this.threadContext.get();
+            final TriConsumer<HttpPreRequest, Channel, ActionListener<ValidationResultContext>> authenticateMessage = (
+                httpRequest,
+                channel,
+                listener) -> {
+                // step 1: stash the current thread's context and wrap the authentication result listener for context-preserving because
+                // authentication modifies the thread context and the modifications must NOT be visible for this thread's yet
+                // (authentication is a very thread-context dependent, as it looks into it for the credentials, and outputs to it
+                // the resulting principal - such an authenticated context must be instated only while executing the request handler)
+                var contextPreservingListener = new ContextPreservingActionListener<>(
+                    threadContext.wrapRestorable(threadContext.newStoredContext()),
+                    listener
+                );
+                try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                    // step 2: populate the thread context with credentials and any other HTTP request header values
+                    perRequestThreadContext.accept(httpRequest, threadContext);
+                    populateClientCertificate.accept(channel, threadContext);
+                    RemoteHostHeader.process(channel, threadContext);
+                    // step 3: run authentication on the properly prepared thread-context
+                    authenticationService.authenticate(httpRequest, ActionListener.wrap(authentication -> {
+                        if (authentication == null) {
+                            logger.trace("No authentication available for HTTP request [{}]", httpRequest.uri());
+                        } else {
+                            logger.trace("Authenticated HTTP request [{}] as {}", httpRequest.uri(), authentication);
+                        }
+                        // step 4: "capture" the *authenticated* thread context
+                        ThreadContext.StoredContext authenticatedContext = threadContext.newStoredContext();
+                        // step 5: restore the original thread-context for the listener and pass in the captured and authenticated context
+                        // the passed-in authenticated context will be instated for the request handler
+                        contextPreservingListener.onResponse(authenticatedContext::restore);
```

Comment:
I think this is the source of alot of my confusion... we use a context preserving listener, but don't really preserve the context for anything meaningful and instead explicitly restore the authenticated context ?   Is there any way to NOT play with the thread context so much ? Niavely, it seems that we just need to stash before authC and populate the context (won't the rest of it happen under the same call stack ?)

Reply:
> I think this is the source of alot of my confusion... we use a context preserving listener, but don't really preserve the context for anything meaningful and instead explicitly restore the authenticated context ?

The intent is that we call the original listener's `#onResponse` method under the context as it was at the time the `contextPreservingListener` was constructed (i.e. the context of a netty worker thread), rather than use the context at the time `#onResponse` is called (an authenticated context, based on the HTTP request headers).
The argument to the `#onResponse` method is a lambda that restores the authenticated context. The authenticated context is restored after all netty layers have passed and we're at the point to run the request handler method.

Reply:
> Is there any way to NOT play with the thread context so much ? Niavely, it seems that we just need to stash before authC and populate the context (won't the rest of it happen under the same call stack ?)

If at this point we call the `listener#onResponse` (rather than the `contextPreservingListener#onResponse`), the call stack will indeed continue as "authenticated", but this is not useful and potentially dangerous (it's "leaking" authn context to code that doesn't use it, i.e. netty, but has the potential to reach ES code that does).
The more cautious intent is that only code that runs under the `RestController#dispatchRequest` will have an authenticated context.